// markmap-render/dist/index.js
"use strict";
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const markmapCommon = require("markmap-common");
const template = '<!DOCTYPE html>\n<html>\n<head>\n<meta charset="UTF-8">\n<meta name="viewport" content="width=device-width, initial-scale=1.0">\n<meta http-equiv="X-UA-Compatible" content="ie=edge">\n<title>Markmap</title>\n<style>\n* {\n  margin: 0;\n  padding: 0;\n}\n#mindmap {\n  display: block;\n  width: 100vw;\n  height: 100vh;\n}\n</style>\n<!--CSS-->\n</head>\n<body>\n<svg id="mindmap"></svg>\n<!--JS-->\n</body>\n</html>\n';
const baseJsPaths = [
  `d3@${"7.8.5"}/dist/d3.min.js`,
  `markmap-view@${"0.17.2"}/dist/browser/index.js`
];
function fillTemplate(root, assets, extra) {
  extra = {
    ...extra
  };
  const urlBuilder = extra.urlBuilder ?? markmapCommon.urlBuilder;
  extra.baseJs ?? (extra.baseJs = baseJsPaths.map((path) => urlBuilder.getFullUrl(path)).map((path) => markmapCommon.buildJSItem(path)));
  const { scripts, styles } = assets;
  const cssList = [...styles ? markmapCommon.persistCSS(styles) : []];
  const context = {
    getMarkmap: () => window.markmap,
    getOptions: extra.getOptions,
    jsonOptions: extra.jsonOptions,
    root
  };
  const jsList = [
    ...markmapCommon.persistJS(
      [
        ...extra.baseJs,
        ...scripts || [],
        {
          type: "iife",
          data: {
            fn: (getMarkmap, getOptions, root2, jsonOptions) => {
              const markmap = getMarkmap();
              window.mm = markmap.Markmap.create(
                "svg#mindmap",
                (getOptions || markmap.deriveOptions)(jsonOptions),
                root2
              );
            },
            getParams: ({ getMarkmap, getOptions, root: root2, jsonOptions }) => {
              return [getMarkmap, getOptions, root2, jsonOptions];
            }
          }
        }
      ],
      context
    )
  ];
  const html = template.replace("<!--CSS-->", () => cssList.join("")).replace("<!--JS-->", () => jsList.join(""));
  return html;
}
exports.baseJsPaths = baseJsPaths;
exports.fillTemplate = fillTemplate;
exports.template = template;
